### PR TITLE
Feat/tracking changes

### PR DESCRIPTION
**Pull request by thestonefox**

### CODE REVIEW

File: Runtime/Tracking/Velocity/ComponentTrackerProxy.cs [hunk 1]
```diff
@@ -27,20 +26,11 @@ public class ComponentTrackerProxy : VelocityTracker
...
-        /// <param name="proxySource">The <see cref="GameObject"/> that contains a <see cref="VelocityTracker"/>.</param>
-        public virtual void SetProxySource(GameObject proxySource)
-        {
-            ProxySource = proxySource == null ? null : proxySource.GetComponent<Component>();
+            return base.IsActive() && cachedVelocityTracker != null && cachedVelocityTracker.isActiveAndEnabled;
         }
 
         /// <summary>
-        /// Clears the existing proxy source.
+        /// Clears the existing <see cref="ProxySource"/>.
```
Comment (bddckr) [inline Runtime/Tracking/Velocity/ComponentTrackerProxy.cs]: We can remove this method now! Malimbe's injected `Clear` method just sets the property to `null` **by using the property setter**, thus the `OnAfterProxySourceChange` below is called and handles this already.

File: Runtime/Tracking/Modification/ComponentEnabledStateModifier.cs [hunk 2]
```diff
@@ -18,29 +17,30 @@ public class ComponentEnabledStateModifier : MonoBehaviour
         /// <summary>
         /// The <see cref="Object"/> types to manage the enabled state on.
         /// </summary>
-        [TypePicker(typeof(Object)), DocumentedByXml]
-        public List<SerializableType> types = new List<SerializableType>();
+        [Serialized]
+        [field: DocumentedByXml, TypePicker(typeof(Object))]
```
[the remark this answers is not in the document]
Comment (thestonefox) [reply Runtime/Tracking/Modification/ComponentEnabledStateModifier.cs]: This is just a brain fart. I don't need the TypePicker on the ObsList. I just need to have the ObsList. The SerializableTypeObservableList already implements the TypePicker.

tldr; this should simply be:
```
[Serialized]
[field: DocumentedByXml]
```

File: Runtime/Tracking/Modification/GameObjectStateSwitcher.cs [hunk 3]
```diff
@@ -42,8 +46,13 @@ public class GameObjectStateSwitcher : MonoBehaviour
         [RequiresBehaviourState]
         public virtual void SwitchNext()
         {
+            if (Targets == null)
```
[the remark this answers is not in the document]
Comment (thestonefox) [reply Runtime/Tracking/Modification/GameObjectStateSwitcher.cs]: why even do the Targets == null check in OnEnable? If it's required and not set then it should always error when the script is running, no?

[the remark this answers is not in the document]
Comment (bddckr) [reply Runtime/Tracking/Modification/GameObjectStateSwitcher.cs]: Yeah but remember our tests hitting this issue? We said we want to support null stuff for folks who AddComponent and those will have a hard time setting the field before Unuty already called OnEnable.
tl;dr: Only support null for required fields when called from "startup" methods for AddComponent usages.

File: Runtime/Tracking/Modification/ComponentEnabledStateModifier.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bddckr) [reply Runtime/Tracking/Modification/ComponentEnabledStateModifier.cs]: I think there still is an important difference: This was telling the picker to use Object while our SerializableypeObservableList is configured to do a Component pick...

[the remark this answers is not in the document]
Comment (thestonefox) [reply Runtime/Tracking/Modification/ComponentEnabledStateModifier.cs]: What's the difference in this case between requiring to pick an Object and requiring to pick a component?

Objects aren't based off of Behaviour right, therefore don't have an enabled state?

Components are and therefore do have an enabled state.

This class is toggling the enabled state on Components. Perhaps Object in the first place was wrong?

[the remark this answers is not in the document]
Comment (bddckr) [reply Runtime/Tracking/Modification/ComponentEnabledStateModifier.cs]: Oh yeah it definitely was wrong. This class wants the enabled property to set. This means it needs a Behaviour (not MonoBehaviour) picker actually. The inheritance hierarchy is a little weird: Behaviour defines the enabled property. Behaviour inherits from Component, but again it's Behaviour actually defining it.

So it still looks like we need two serializable type ObsLists?!

File: Runtime/Tracking/Modification/GameObjectStateSwitcher.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (thestonefox) [reply Runtime/Tracking/Modification/GameObjectStateSwitcher.cs]: I dont think `Switch On Enable` should be a thing at all.

Again if you wanna do that then use the Behaviour Enabled Observer and call `SwitchTo(int)`

Perhaps then also `SwitchTo(int index)` should be `SwitchTo(int index = StartIndex)` so you can simply call `SwitchTo()`

[the remark this answers is not in the document]
Comment (bddckr) [reply Runtime/Tracking/Modification/GameObjectStateSwitcher.cs]: Good point. It's probably worth naming it CurrentIndex at that point and support that at any time, like the index on ObsList works!

Do UnityEvents allow calling methods with a default parameter like that? I think we tried that in the past and had to offer a new method taking no arguments instead.

File: Runtime/Tracking/Modification/ComponentEnabledStateModifier.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (thestonefox) [reply Runtime/Tracking/Modification/ComponentEnabledStateModifier.cs]: The current SerializedType ObsList seems to give us every component that you can enable/disable

(before the property type was SerializedType too and we never hit any issues)

Or are we saying we're going to get a list of Components and many won't be Behaviours therefore you can pick things that will crash it?

[the remark this answers is not in the document]
Comment (bddckr) [reply Runtime/Tracking/Modification/ComponentEnabledStateModifier.cs]: Yeah the latter. The type isn't guaranteed to offer the enabled property. A user could pick something that doesn't have it like a transform I think.

File: Runtime/Tracking/Modification/GameObjectStateSwitcher.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (thestonefox) [reply Runtime/Tracking/Modification/GameObjectStateSwitcher.cs]: Actually the issue is you can't use a variable for the default value

`SwitchTo(int index = MyVar)` won't work, has to be a const

So we'll probably need a `SwitchToStartIndex()` anyway

File: Runtime/Tracking/Modification/ComponentEnabledStateModifier.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (thestonefox) [reply Runtime/Tracking/Modification/ComponentEnabledStateModifier.cs]: I'm going to create a `SerializedTypeBehaviourObservableList` in this commit but leave `SerializableTypeObservableList` unchanged (then i'll add another commit at the end that renames it)
